lib/default.nix

```diff
@@ -393,4 +393,96 @@ in {
   # one in the list will be used.
   combineFlakes = sep: prefixAndFlakes: builtins.foldl' addFlakes {}
     (lib.mapAttrsToList (prefix: flake: prefixFlake prefix sep flake) prefixAndFlakes);
+
+  mkFlake = project: packages: coverageProjectModule:
+    let
+      packageNames = project: builtins.attrNames (project.args.flake.packages project.hsPkgs);
+      checkedProject = project.appendModule { checkMaterialization = true; };
+      coverageProject = project.appendModule [
+        coverageProjectModule
+        {
+          modules = [{
+            packages = lib.genAttrs (packageNames project)
+              (_: { doCoverage = lib.mkDefault true; });
+          }];
+        }
+      ];
+      checks = builtins.listToAttrs (
+        lib.concatMap (packageName:
+          let package = project.hsPkgs.${packageName};
+          in lib.mapAttrsToList (n: v:
+              { name = "${packageName}:test:${n}"; value = v; })
+            (lib.filterAttrs (_: v: lib.isDerivation v) (package.checks))
+        ) (packageNames project));
+      ciJobs = {
+          # Run all the tests
+          inherit checks;
+          # Build tools and cache tools needed for the project
+          inherit (project) roots;
+          # Also build and cache any tools in the `devShell`
+          devShell = project.shell;
+          # Run HPC on the tests
+          coverage = builtins.listToAttrs (lib.concatMap (packageName: [{
+            name = packageName;
+            value = coverageProject.hsPkgs.${packageName}.coverageReport;
+          }]) (packageNames coverageProject));
+        }
+        # Build the plan-nix and check it if materialized
+        // lib.optionalAttrs (checkedProject ? plan-nix) {
+          plan-nix = checkedProject.plan-nix;
```

Comment:
Actually, do we want to build this unconditionally from the main project also? Isn't there something funny about how `haskell.nix` fails when cabal solving fails and the error ends up as the result of this derivation? I forget

Reply:
I think the problem was that the error was reported in hydra's eval log (for which there was only one per jobset).  I think it might still be a problem, but the eval error logs work better now.  See https://github.com/input-output-hk/haskell.nix/pull/1744